Initial Exporter Implementation

ewr opened the pull request:
Here's my initial implementation.

Currently provides service health per node and number of Consul servers in the cluster.

brian-brazil commented on consul_exporter.go:
```diff
@@ -0,0 +1,216 @@
+		}),
+
+		serviceMetrics: map[string]*prometheus.GaugeVec{
+			"nodes_healthy":   newServiceMetric("nodes", "Number of nodes", prometheus.Labels{"healthy": "healthy"}),
```
Generally it's easier to work with to have total and healthy as separate metrics

ewr replied:
So `nodes_total`, and then `nodes` with the label-split for healthy and unhealthy? I wasn't sure the best way to think of this, where basically "healthy" or "unhealthy" is my value. 

I know I want to be able to get counts for total, healthy and unhealthy, overall or by service, and then be able to query unhealthy service nodes `by (service,node)` to see the actual failing combo.

brian-brazil replied:
Hmm, this is a boolean. What you want then is a single metric called `node_healthy` with (service, node) as labels and a value of 0 or 1. That should let you do everything that you'd want to do with that data without getting into complex label matching.

ewr replied:
Ah... I missed in the query logic that I could do `count(consul_service_nodes_healthy == 0)`, etc. That's much better. I'll change that.

brian-brazil replied:
That's not quite correct as it'll return an empty vector if everything is healthy. You want `sum by (service)(1 - consul_service_nodes_healthy)`

brian-brazil commented on consul_exporter.go:
```diff
@@ -0,0 +1,216 @@
+		}),
+		totalQueries: prometheus.NewCounter(prometheus.CounterOpts{
+			Namespace: namespace,
+			Name:      "exporter_total_queries",
```
this would be queries_total

I'm not sure there's much point tracking this

ewr replied:
I don't know of one. I copied it from the haproxy exporter. I'll remove it unless someone says to keep it.

juliusv replied:
Actually, this confused me. I thought it meant the number of client queries against Consul, not how often this exporter has queried Consul. Yeah, I think this can be removed. But stats about client queries could actually be useful, if they're obtainable.

juliusv commented on .gitignore:
```diff
@@ -0,0 +1,8 @@
+.build
+bin/consul_exporter*
```
Could you change this to use the `Makefile.COMMON` from https://github.com/prometheus/utils/blob/master/Makefile.COMMON and then this will actually just be called `consul_expoter`. I think the `bin/...` and `prometheus-consul_exporter` lines can go away then.

See https://github.com/prometheus/mesos_exporter/blob/master/Makefile for an example of how to use the common Makefile (which you don't need to modify).

If you prefer, I can also do that in a followup PR, no problem.

juliusv commented on consul_exporter.go:
```diff
@@ -0,0 +1,216 @@
+
+import (
+	"flag"
+	_ "fmt"
```
Hum, lots of unused imports that can be removed?

But I'd recommend adding:

    _ "net/http/pprof"

...so that it's easier to debug in production.

ewr replied:
I started from the haproxy_exporter importers, then just started commenting things out to see what I needed. The missing step was going back to actually delete the ones I didn't.

juliusv commented on consul_exporter.go:
```diff
@@ -0,0 +1,216 @@
+	e.mutex.Lock() // To protect metrics from concurrent collects.
+	defer e.mutex.Unlock()
+
+	// reset metrics
```
Nitpick: we use sentence-style for comments everywhere, a la "Reset metrics.". Especially for multi-line comments like one of the comments you have below, that improves readability/cleanliness.

ewr commented:
I haven't done the Makefile yet, but I think this commit cleans up the service metric, and adds a few new ones:

    * Register `consul_service_nodes_total` for each service, and
      `consul_service_nodes_healthy` for each service/node combo. Use value of
      `1` for healthy, `0` for unhealthy instead of using labels.
    * Add `consul_nodes` as a counter for the number of nodes registered with
      Consul (regardless of services)
    * Add `consul_services` as a counter for the number of registered services

I also removed commented-out includes and fixed up at least some comments to be sentence-style.

grobie commented on consul_exporter.go:
```diff
@@ -0,0 +1,227 @@
+		passing := 1
+
+		for _, hc := range entry.Checks {
+			if hc.Status != "passing" {
```
We should check against the constant consul provides instead of using the string here.

```go
import consul "github.com/hashicorp/consul/consul/structs"
if check.Status != consul.HealthPassing {
  ...
}
```

grobie commented on consul_exporter.go:
```diff
@@ -0,0 +1,227 @@
+
+		for _, hc := range entry.Checks {
+			if hc.Status != "passing" {
+				passing = 0
```
Break out of the loop here.

grobie commented:
Some initial questions and thoughts @ewr:

* Any objections to export all metrics consul already provides (in an automated way)? https://consul.io/docs/commands/info.html I can port over our existing code for that.
* Can we use the metric names consul already provides, so `consul_agent_services` instead of `consul_services`? This should make it easier for people to match the prometheus metrics to what consul documents.
* Consul also already provides a metrics for LAN members, which is what you mean with `consul_nodes` right? So what about `consul_serf_lan_members`?

ewr commented:
Pulling in your existing Consul stats code sounds great, but it looks like it adds a dependency that the consul binary be available locally. Using the API package, we can connect to a local agent or to a remote server without needing the binary. Maybe we make it optional with a flag?

If the `serf_lan_members` in there already equals nodes, then there's no reason to pull that separately. We're only using the count, not the names. We may need to preserve the existing approach to fetch it if we're not pulling info from a local binary.

I'm not sure on the agent terminology... I think it's a perspective question. In the current implementation, we're querying at one point for all services registered in Consul. `agent_services` would be the services registered on the local agent, whereas `services` is the count of all services known across the DC. I think that (and specifically nodes on services rather than agents) is in keeping with the Consul API around `catalog/services`.

In your case with glimpse, are you pulling stats from every agent, or just globally on the cluster?

grobie commented:
Yeah, I don't like the cmd execution path either, I'll probably contribute a change to provide the stats as JSON via HTTP. So long a flag sounds fine.
 
Maybe let's keep the existing approach for now until we can fetch the stats via HTTP, but rename to `serf_lan_members` already? Then people won't need to rename later.

You're right, `agent_services` would be a different number, good point. What do you think about using a `catalog` prefix for all the the metrics you're introducing here? That should help to both avoid any naming clashes with consul's metrics as well as to make clear what the metric is about in opposite to `agent_services`?

Yes, we're scraping all glimpse-agents (in the magnitude of thousands).

ewr commented:
@grobie: I've added the `catalog_` prefix and renamed `nodes` and `servers`. I think there's probably going to be a point where it makes sense for this exporter to support flags turning on and off catalog and agent-specific features. You might have one exporter grabbing the catalog, but then a bunch of them doing per-node agent exports.

@brian-brazil: I added a `consul_catalog_service_healthy` boolean, which is true if all checks for the service are passing. That's something that I know you could figure out via looking at the `consul_catalog_service_node_healthy` results for the service, so I'm open to suggestions on when providing a roll-up like that makes sense vs when it should just be something that gets computed if needed.

I realized this morning that all of the catalog metrics need a `dc` label, but it's not super-clear to me in the Consul API docs how I would grab that from the agent we connect to. I don't run a multi-DC setup, so I don't have experience there.

ewr commented:
Is there a way that cross-compiling is intended to work with the common Makefile? It seems like just changing these two lines to conditional definitions would do it:

```
GOOS   ?= $(shell uname | tr A-Z a-z)
GOARCH ?= $(subst x86_64,amd64,$(patsubst i%86,386,$(shell uname -m)))
```

For instance:

```
~/projects/scpr/consul_exporter: GOOS=linux GOARCH=amd64 make archive
GOROOT=/usr/local/Cellar/go/1.4.2/libexec GOPATH=/Users/eric/projects/scpr/consul_exporter/.build/gopath /usr/local/Cellar/go/1.4.2/libexec/bin/go build  -o consul_exporter
tar -czf consul_exporter-0.1.0.linux-amd64.tar.gz consul_exporter
```

grobie commented:
Makes sense to change that. Open a PR in prometheus/utils please.
https://github.com/prometheus/utils/blob/master/Makefile.COMMON

On Thu, Mar 19, 2015 at 10:52 AM, Eric Richardson <notifications@github.com>
wrote:

> Is there a way that cross-compiling is intended to work with the common
> Makefile? It seems like just changing these two lines to conditional
> definitions would do it:
>
> GOOS   ?= $(shell uname | tr A-Z a-z)
> GOARCH ?= $(subst x86_64,amd64,$(patsubst i%86,386,$(shell uname -m)))
>
> For instance:
>
> ~/projects/scpr/consul_exporter: GOOS=linux GOARCH=amd64 make archive
> GOROOT=/usr/local/Cellar/go/1.4.2/libexec GOPATH=/Users/eric/projects/scpr/consul_exporter/.build/gopath /usr/local/Cellar/go/1.4.2/libexec/bin/go build  -o consul_exporter
> tar -czf consul_exporter-0.1.0.linux-amd64.tar.gz consul_exporter
>
> --
> Reply to this email directly or view it on GitHub
> <https://github.com/prometheus/consul_exporter/pull/1#issuecomment-83614593>
> .
>

brian-brazil commented:
> I'm open to suggestions on when providing a roll-up like that makes sense vs when it should just be something that gets computed if needed.

The general rule is that you only provide something like that if you're running into performance issues.

In this particular instance, I'd be wary of reporting something as healthy only if every single node was working as a single node being unhealthy doesn't necessarily mean that the service as a whole is unhealthy. That's the sort of thing that's service and site specific, so I wouldn't build it into a generic exporter.

ewr commented:
That's fair.

What would the query be to get that same result based on the service_node_healthy metrics? As you mentioned before, `sum by(service)(1 - consul_catalog_service_node_healthy)` gives you a count of unhealthy nodes per service, but I'm struggling to figure out a counterpart that would compare the count of nodes for a service with `consul_catalog_service_node_healthy == 1` against `consul_catalog_service_nodes` to see whether everything is healthy.

brian-brazil commented:
`min(consul_catalog_service_node_healthy ) == 0`

ewr commented:
Thanks. Pulled that metric and added a start to README documentation on useful ways to pull out the data.

brian-brazil commented:
One thing to consider is adding a default console template to prometheus itself, so that users get a dashboard out of the box if they're scraping a consul exporter. See https://github.com/prometheus/prometheus/tree/master/consoles

grobie commented on Makefile:
```diff
@@ -0,0 +1,17 @@
+VERSION  := 0.1.0
+TARGET   := consul_exporter
+
+include Makefile.COMMON
```
Newline at the end of files please. I saw a few more files where it's also missing.

grobie commented on README.md:
````diff
@@ -1,3 +1,30 @@
+To run it:
+
+```bash
+go run haproxy_exporter [flags]
````
```bash
make
./consul_exporter [flags]
```

grobie commented on README.md:
````diff
@@ -1,3 +1,30 @@
+
+Help on flags:
+```bash
+go run haproxy_exporter --help
````
```
./consul_exporter --help
```

And an empty line after `Help on flags:` please.

grobie commented on consul_exporter.go:
```diff
@@ -0,0 +1,235 @@
+	var (
+		listenAddress = flag.String("web.listen-address", ":9105", "Address to listen on for web interface and telemetry.")
+		metricsPath   = flag.String("web.telemetry-path", "/metrics", "Path under which to expose metrics.")
+		consulServer  = flag.String("consul.server", "localhost:8500", "URI for Consul server.")
```
What do you think about `HTTP API address of a Consul server.`?

grobie commented:
A few very minor comments, looks good! LGTM.

:+1:

juliusv commented on consul_exporter.go:
```diff
@@ -0,0 +1,235 @@
+
+func main() {
+	var (
+		listenAddress = flag.String("web.listen-address", ":9105", "Address to listen on for web interface and telemetry.")
```
Oops, this port has been taken by two new exporters in the meantime. Could you change it to 9107? (https://github.com/prometheus/prometheus/wiki/Default-port-allocations)

ewr commented:
Thanks @grobie and @juliusv. I think that commit should address your comments.

ewr commented:
I went ahead and removed the unused flags. I do want to circle back and implement the locks one soon, but there's no reason for it to be there until I do.

juliusv commented:
Woot, thanks!
